lib/sqlalchemy/sql/functions.py

```diff
@@ -49,7 +56,38 @@ def register_function(identifier, fn, package="_default"):
 
     """
     reg = _registry[package]
-    reg[identifier] = fn
+    raw_identifier = identifier
+    identifier = identifier.lower()
+
+    # Check if a function with the same lowercase identifier is registered.
+    if identifier in _case_sensitive_reg[package]:
+        if (
+            raw_identifier not in _case_sensitive_reg[package][identifier]
+        ):
+            util.warn_deprecated(
+                "GenericFunction(s) '{}' are already registered with "
+                "different letter cases and might interact with {}.".format(
+                    _case_sensitive_reg[package][identifier],
+                    raw_identifier))
+            # If a function with the same lowercase identifier is registered,
+            # then these 2 functions are considered as case-sensitive.
+            if len(_case_sensitive_reg[package][identifier]) == 1:
+                old_fn = reg[identifier]
+                del reg[identifier]
+                reg[_case_sensitive_reg[package][identifier][0]] = old_fn
+        else:
+            warnings.warn(
+                "The GenericFunction '{}' is already registered and "
+                "is going to be overriden.".format(identifier),
+                sa_exc.SAWarning)
+
+        reg[raw_identifier] = fn
+    else:
+        reg[identifier] = fn
+
+    # Add the raw_identifier to the case-sensitive registry
```

Comment:
**mike bayer** (zzzeek) wrote:

so if this is a dict, it becomes:

    _case_sensitive_reg[[package][identifier][raw_identifier] = fn

Reply:
Ok